go-backend/db/user.go

```diff
@@ -2,21 +2,135 @@ package db
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting user's organization from database by email and id " + email + " " + string(user.OrgID))
+	}
+	return
+}
+
+// GetUserByID - Given the database ID for that user, look them up.
+func (s *pgStore) GetUserByID(ctx context.Context, id int) (user User, err error) {
+	err = s.db.Get(&user, `SELECT * FROM users WHERE id=$1 LIMIT 1`, id)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting user from database by id " + string(id))
+		return
+	}
+	// Populate the user's organization data
+	org := Organization{}
+	err = s.db.Get(&org, `SELECT * FROM organizations WHERE id=$1`, user.OrgID)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting user's organization from database by user id and org id " + string(id) + " " + string(user.OrgID))
+	}
+	return
+}
+
+// ListUsers - retrieves all users from the database. Could be a very large result set...
 func (s *pgStore) ListUsers(ctx context.Context) (users []User, err error) {
 	err = s.db.Select(&users, "SELECT * FROM users ORDER BY name ASC")
 	if err != nil {
 		logger.WithField("err", err.Error()).Error("Error listing users")
+	}
+	return
+}
+
+// CreateNewUser - creates a new user in the database
+func (s *pgStore) CreateNewUser(ctx context.Context, u User) (newUser User, err error) {
+	// First, make sure they're not already in the database because if they are, we can just return that user
+	newUser, err = s.GetUserByEmail(ctx, u.Email)
+	if err == nil {
+		// If there's already a user here err will be nil, so newUser is populated. Just return, don't
+		// bother with all the rest of this function. No need to log an error that doesn't exist, either.
 		return
 	}
 
+	// If we made it this far, they don't appear to be in the database yet.
+	q := `INSERT INTO users (
+    id, name, org_id, email, display_name, profile_image_url, soft_delete, role_id, hi5_quota_balance,
+    soft_delete_by, soft_delete_on, created_at
+  ) VALUES (
+    DEFAULT, :name, :org_id, :email, :display_name, :profile_image_url, FALSE, :role_id, :hi5_quota_balance,
+    0, NULL, :created_at
+	)`
+
+	// Set the created_at time property on u so that it doesn't default to some weird value over 2000 years ago
+	u.CreatedAt = time.Now().UTC()
+
+	// Set the user's role; we're going to start by looking up the role for "Employee" (automatically created by
+	// database migrations so we know it's there), then assign the user's RoleID to that role's ID.
+	r, _ := s.GetRoleByName(ctx, "Employee") // TODO: What should we do here if there's no "Employee" role?
+	u.RoleID = r.ID
+
+	tx, err := s.db.Beginx() // Use Beginx instead of MustBegin so process doesn't die if there's an error
+	if err != nil {
+		// FAIL: Could not begin database transaction
+		logger.WithField("err", err.Error()).Error("Error beginning user insert transaction in db.CreateNewUser with email " + u.Email)
```

Comment:
is a return required here in case of error?

Reply:
010a355cbaf1bee655f9dbe74fb3fae70dc5fade